src/python/foglamp/core/scheduler_db_services.py

```diff
@@ -0,0 +1,285 @@
+
+    # TODO: Investigate why prepared statement is not working
+    # if schedule_id is None:
+    #     _where_clause = ''
+    # else:
+    #     _where_clause = ' WHERE id = \'$1\''
+    #
+    # query += _where_clause
+    #
+    # stmt = await conn.prepare(query)
+    # if schedule_id is None:
+    #     rows = await stmt.fetch()
+    # else:
+    #     rows = await stmt.fetch(schedule_id)
+
+    if not schedule_id:
+        _where_clause = ''
+    else:
+        _where_clause = ' WHERE id = \'{0}\''
+
+    query += _where_clause
+
+    if schedule_id is not None:
+        query = query.format(schedule_id)
+
+    rows = await conn.fetch(query)
+
+    columns = (
+        'id',
+        'process_name',
+        'schedule_name',
+        'schedule_type',
+        'schedule_interval',
+        'schedule_time',
+        'schedule_day',
+        'exclusive'
+    )
+
+    results = []
+    for row in rows:
+        results.append(dict(zip(columns, row)))
+
+    await conn.close()
+
+    return results
+
+async def update_schedule(payload):
+    pass
+
+async def delete_schedule(payload):
+    pass
+
+
+async def create_task(payload):
+    pass
+
+
+async def read_task(task_id=None, state=None, name=None):
+    conn = await asyncpg.connect(dsn=__DSN)
+    query = '''
+        SELECT
+            id::"varchar",
+            process_name,
+            state,
+            start_time::"varchar",
+            end_time::"varchar",
+            reason,
+            pid,
+            exit_code
+        from tasks
+    '''
+
+    # TODO: Investigate why prepared statement is not working
+    # if not task_id:
+    #     _where_clause = ''
+    # else:
+    #     _where_clause = ' WHERE id = \'$1\''
+    #
+    # query += _where_clause
+    #
+    # if task_id is not None:
+    #     stmt = await conn.prepare(query)
+    #     rows = await stmt.fetch(task_id)
+    # else:
+    #     if state is None and name is None:
+    #         _where_clause = ''
+    #     elif state is None and name is not None:
+    #         _where_clause = ' WHERE process_name = \'$1\''
+    #     elif state is not None and name is None:
+    #         _where_clause = ' WHERE state = \'$1\''
+    #     else:
+    #         _where_clause = ' WHERE process_name = \'$1\' and state = \'$2\''
+    #
+    #     query += _where_clause
+    #
+    #     stmt = await conn.prepare(query)
+    #     if state is None and name is None:
+    #         rows = await stmt.fetch()
+    #     elif state is None and name is not None:
+    #         rows = await stmt.fetch(name)
+    #     elif state is not None and name is None:
+    #         rows = await stmt.fetch(state)
+    #     else:
+    #         rows = await stmt.fetch(state, name)
+
+    if not task_id:
+        _where_clause = ''
+    else:
+        _where_clause = ' WHERE id = \'{0}\''
+
+    query += _where_clause
+
+    if task_id is not None:
+        query = query.format(task_id)
+    else:
+        if state is None and name is None:
```

Comment:
This seems a little cumbersome, you could reduce the number of if clauses buy always have "WHERE 1 = 1", that way for each condition you want to add you simply say " AND <column> = <value)". Doing it this way as you add more filters the number of if .. elif clauses just explodes.